File: code/01_stacks_qc_snp_calling/run_sample_missingness_outliers.R
#!/usr/bin/env Rscript
options(stringsAsFactors = FALSE)

base_dir <- "/Users/gspellman/Trowbridgii_analyses/Stacks analysis of ddRAD data/stacks_refmap_sorex_excl_MVZ216210_2026-02-26_clean"
down_dir <- file.path(base_dir, "downstream_analysis_2026-02-23")
vcf_file <- file.path(base_dir, "populations.snps.vcf")
pop_file <- file.path(base_dir, "popmap.tsv")
tab_dir <- file.path(down_dir, "tables")
fig_dir <- file.path(down_dir, "figures")
dir.create(tab_dir, recursive = TRUE, showWarnings = FALSE)
dir.create(fig_dir, recursive = TRUE, showWarnings = FALSE)

is_missing_gt <- function(x) {
  gt <- sub(":.*", "", x)
  gt <- gsub("\\|", "/", gt)
  gt %in% c("./.", ".")
}

con <- file(vcf_file, "r")
on.exit(close(con), add = TRUE)

samples <- NULL
missing_counts <- NULL
total_sites <- 0L

repeat {
  ln <- readLines(con, n = 1)
  if (!length(ln)) break
  if (startsWith(ln, "#CHROM")) {
    hdr <- strsplit(ln, "\t", fixed = TRUE)[[1]]
    samples <- hdr[10:length(hdr)]
    missing_counts <- integer(length(samples))
    next
  }
  if (startsWith(ln, "#")) next
  p <- strsplit(ln, "\t", fixed = TRUE)[[1]]
  if (length(p) < 10) next
  g <- p[10:length(p)]
  total_sites <- total_sites + 1L
  miss <- vapply(g, is_missing_gt, logical(1))
  missing_counts <- missing_counts + as.integer(miss)
}

if (is.null(samples) || total_sites == 0) stop("Failed to parse VCF for samples/sites")

miss_prop <- missing_counts / total_sites
callrate <- 1 - miss_prop

popmap <- read.table(pop_file, header = TRUE, sep = "\t", check.names = FALSE)
colnames(popmap) <- c("Sample", "Population")

df <- data.frame(
  Sample = samples,
  N_sites = total_sites,
  N_missing = missing_counts,
  Missingness = miss_prop,
  Callrate = callrate,
  stringsAsFactors = FALSE
)
df <- merge(df, popmap, by = "Sample", all.x = TRUE)

q1 <- as.numeric(stats::quantile(df$Missingness, 0.25, na.rm = TRUE))
q3 <- as.numeric(stats::quantile(df$Missingness, 0.75, na.rm = TRUE))
iqr <- q3 - q1
lower <- max(0, q1 - 1.5 * iqr)
upper <- q3 + 1.5 * iqr

df$Outlier_IQR <- (df$Missingness < lower) | (df$Missingness > upper)
mu <- mean(df$Missingness, na.rm = TRUE)
sd_m <- stats::sd(df$Missingness, na.rm = TRUE)
if (is.finite(sd_m) && sd_m > 0) {
  df$Missingness_Z <- (df$Missingness - mu) / sd_m
} else {
  df$Missingness_Z <- 0
}
df$Outlier_Z3 <- abs(df$Missingness_Z) > 3

df <- df[order(df$Missingness, decreasing = TRUE), , drop = FALSE]

full_tab <- file.path(tab_dir, "sample_missingness_full_dataset.tsv")
out_tab <- file.path(tab_dir, "sample_missingness_outliers.tsv")
sum_txt <- file.path(tab_dir, "sample_missingness_summary.txt")
write.table(df, full_tab, sep = "\t", quote = FALSE, row.names = FALSE)
write.table(df[df$Outlier_IQR | df$Outlier_Z3, , drop = FALSE], out_tab, sep = "\t", quote = FALSE, row.names = FALSE)

cat(
  "Per-sample missingness summary (full dataset)\n",
  sprintf("VCF: %s\n", vcf_file),
  sprintf("Samples: %d\n", nrow(df)),
  sprintf("Sites: %d\n", total_sites),
  sprintf("Mean missingness: %.6f\n", mean(df$Missingness)),
  sprintf("Median missingness: %.6f\n", median(df$Missingness)),
  sprintf("Q1: %.6f\n", q1),
  sprintf("Q3: %.6f\n", q3),
  sprintf("IQR: %.6f\n", iqr),
  sprintf("IQR lower bound: %.6f\n", lower),
  sprintf("IQR upper bound: %.6f\n", upper),
  sprintf("IQR outliers: %d\n", sum(df$Outlier_IQR)),
  sprintf("|Z| > 3 outliers: %d\n", sum(df$Outlier_Z3)),
  file = sum_txt
)

hist_png <- file.path(fig_dir, "sample_missingness_histogram_full_dataset.png")
hist_pdf <- file.path(fig_dir, "sample_missingness_histogram_full_dataset.pdf")

plot_hist <- function(dev_fun, file) {
  dev_fun(file)
  par(mar = c(5, 5, 3, 1))
  vals <- 100 * df$Missingness
  h <- hist(vals, breaks = 20, col = "#4C78A8", border = "white",
            main = "Per-sample missingness distribution (full SNP dataset)",
            xlab = "Per-sample missingness (%)", ylab = "Number of samples")
  abline(v = 100 * upper, col = "#D62728", lwd = 2, lty = 2)
  abline(v = 100 * median(df$Missingness), col = "#2F2F2F", lwd = 1.5)
  if (sum(df$Outlier_IQR) > 0) {
    for (x in 100 * df$Missingness[df$Outlier_IQR]) {
      abline(v = x, col = "#E45756", lwd = 1, lty = 1)
    }
  }
  legend("topright",
         legend = c(sprintf("Median = %.2f%%", 100 * median(df$Missingness)),
                    sprintf("IQR upper = %.2f%%", 100 * upper)),
         lty = c(1, 2), lwd = c(1.5, 2), col = c("#2F2F2F", "#D62728"), bty = "n")
  dev.off()
}

plot_hist(function(f) png(f, width = 3000, height = 2000, res = 320), hist_png)
plot_hist(function(f) pdf(f, width = 9.5, height = 6.2, useDingbats = FALSE), hist_pdf)

cat(sprintf("Wrote:\n%s\n%s\n%s\n%s\n%s\n", full_tab, out_tab, sum_txt, hist_png, hist_pdf))

File: code/01_stacks_qc_snp_calling/run_sample_missingness_outliers.py
#!/usr/bin/env python3
import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

BASE = "/Users/gspellman/Trowbridgii_analyses/Stacks analysis of ddRAD data/stacks_refmap_sorex_excl_MVZ216210_2026-02-26_clean"
DOWN = os.path.join(BASE, "downstream_analysis_2026-02-23")
VCF = os.path.join(BASE, "populations.snps.vcf")
POPMAP = os.path.join(BASE, "popmap.tsv")
TAB = os.path.join(DOWN, "tables")
FIG = os.path.join(DOWN, "figures")
os.makedirs(TAB, exist_ok=True)
os.makedirs(FIG, exist_ok=True)


def is_missing(gt_field: str) -> bool:
    gt = gt_field.split(":", 1)[0]
    gt = gt.replace("|", "/")
    return gt in {"./.", ".", "./."}


# Parse VCF and count missing genotypes per sample
samples = None
missing_counts = None
total_sites = 0

with open(VCF, "r") as fh:
    for line in fh:
        if line.startswith("#CHROM"):
            header = line.rstrip("\n").split("\t")
            samples = header[9:]
            missing_counts = np.zeros(len(samples), dtype=np.int64)
            continue
        if line.startswith("#"):
            continue
        parts = line.rstrip("\n").split("\t")
        if len(parts) < 10:
            continue
        gts = parts[9:]
        total_sites += 1
        for i, g in enumerate(gts):
            if is_missing(g):
                missing_counts[i] += 1

if samples is None or total_sites == 0:
    raise RuntimeError("Failed to read samples/sites from VCF")

missing_prop = missing_counts / float(total_sites)
callrate = 1.0 - missing_prop

# Join population labels
pop = pd.read_csv(POPMAP, sep="\t")
pop.columns = ["Sample", "Population"]
df = pd.DataFrame({
    "Sample": samples,
    "N_sites": total_sites,
    "N_missing": missing_counts,
    "Missingness": missing_prop,
    "Callrate": callrate,
})
df = df.merge(pop, on="Sample", how="left")

# Outlier detection (IQR rule)
q1 = float(df["Missingness"].quantile(0.25))
q3 = float(df["Missingness"].quantile(0.75))
iqr = q3 - q1
upper = q3 + 1.5 * iqr
lower = max(0.0, q1 - 1.5 * iqr)

df["Outlier_IQR"] = (df["Missingness"] > upper) | (df["Missingness"] < lower)

# z-score as secondary diagnostic
mu = float(df["Missingness"].mean())
sd = float(df["Missingness"].std(ddof=1))
if sd > 0:
    z = (df["Missingness"] - mu) / sd
else:
    z = np.zeros(len(df))
df["Missingness_Z"] = z
df["Outlier_Z3"] = np.abs(df["Missingness_Z"]) > 3.0

# Write tables
df = df.sort_values("Missingness", ascending=False).reset_index(drop=True)
full_table = os.path.join(TAB, "sample_missingness_full_dataset.tsv")
outlier_table = os.path.join(TAB, "sample_missingness_outliers.tsv")
summary_file = os.path.join(TAB, "sample_missingness_summary.txt")

df.to_csv(full_table, sep="\t", index=False)
df[df["Outlier_IQR"] | df["Outlier_Z3"]].to_csv(outlier_table, sep="\t", index=False)

with open(summary_file, "w") as fh:
    fh.write("Per-sample missingness summary (full dataset)\n")
    fh.write(f"VCF: {VCF}\n")
    fh.write(f"Samples: {len(df)}\n")
    fh.write(f"Sites: {total_sites}\n")
    fh.write(f"Mean missingness: {df['Missingness'].mean():.6f}\n")
    fh.write(f"Median missingness: {df['Missingness'].median():.6f}\n")
    fh.write(f"Q1: {q1:.6f}\n")
    fh.write(f"Q3: {q3:.6f}\n")
    fh.write(f"IQR: {iqr:.6f}\n")
    fh.write(f"IQR lower bound: {lower:.6f}\n")
    fh.write(f"IQR upper bound: {upper:.6f}\n")
    fh.write(f"IQR outliers: {int(df['Outlier_IQR'].sum())}\n")
    fh.write(f"|Z|>3 outliers: {int(df['Outlier_Z3'].sum())}\n")

# Plot histogram
plt.rcParams.update({"font.size": 11})
fig, ax = plt.subplots(figsize=(9, 6))
vals_pct = df["Missingness"].to_numpy() * 100.0
ax.hist(vals_pct, bins=20, color="#4C78A8", edgecolor="white", alpha=0.95)
ax.axvline(upper * 100.0, color="#D62728", linestyle="--", linewidth=1.8, label=f"IQR upper ({upper*100:.2f}%)")
ax.axvline(df["Missingness"].median() * 100.0, color="#2F2F2F", linestyle="-", linewidth=1.2, label=f"Median ({df['Missingness'].median()*100:.2f}%)")

out = df[df["Outlier_IQR"]]
if len(out):
    for _, r in out.iterrows():
        ax.axvline(r["Missingness"] * 100.0, color="#E45756", alpha=0.5, linewidth=1.0)

ax.set_xlabel("Per-sample missingness (%)")
ax.set_ylabel("Number of samples")
ax.set_title("Per-sample missingness distribution (full SNP dataset)")
ax.legend(frameon=True, fontsize=9)
fig.tight_layout()

png = os.path.join(FIG, "sample_missingness_histogram_full_dataset.png")
pdf = os.path.join(FIG, "sample_missingness_histogram_full_dataset.pdf")
fig.savefig(png, dpi=400)
fig.savefig(pdf)
plt.close(fig)

print("Wrote:")
print(full_table)
print(outlier_table)
print(summary_file)
print(png)
print(pdf)
